### PR TITLE
Use verdaccio for the template e2e test

### PR DESCRIPTION
**Pull request by fortmarek**
## Summary

This PR adds [verdaccio](https://github.com/verdaccio/verdaccio) to release packages in the `packages` directory during the E2E test on CI. 

The rationale behind this is the following:
- Firstly, we wanted to push the [monorepo RFC](https://github.com/react-native-community/discussions-and-proposals/pull/480). We hit an issue when renaming the packages to follow the same convention caused by the e2e test using the template to fail. This is because the template installs packages from the live version of npm – and if you just rename a package in a given PR without releasing it, the package understandably can't be installed since it's not published, yet – as you can see [here](https://app.circleci.com/pipelines/github/facebook/react-native/15286/workflows/149df51f-f59b-4eb3-b92c-20c513111f04/jobs/282135?invite=true#step-108-283). 
- Secondly, the current e2e test on `main` does not actually test the latest code of the packages in the `packages` directory as it simply downloads the latest versions from npm. This creates a divide between what's tested and what users should expect when using nightlies or when a new minor is released.

## Changelog

[Internal] [Changed] - Use verdaccio for template e2e test

## Test Plan

`test_js` CI check should pass. Additionally, I have temporarily updated the [PR](https://github.com/facebook/react-native/pull/34572) renaming `assets` to `assets-registry` to include the verdaccio changes – the `test_js` passes there, additionally proving merging this PR will unblock us with the rename PRs.

### CODE REVIEW

File: scripts/run-ci-e2e-tests.js [hunk 1]
```diff
@@ -70,6 +72,25 @@ try {
 
   const REACT_NATIVE_PACKAGE = path.join(ROOT, 'react-native-*.tgz');
 
+  describe('Verdaccio');
+  const verdaccioProcess = child_process.spawn('npx', [
+    'verdaccio',
```
Comment (oblador) [inline scripts/run-ci-e2e-tests.js:77]: I didn't see this added to dependencies, so I think we want to lock the version somehow, you can do it like this:
```suggestion
    'verdaccio@5.15.3',
```

File: scripts/run-ci-e2e-tests.js [hunk 2]
```diff
@@ -156,9 +179,13 @@ try {
...
     describe(`Start Metro, ${SERVER_PID}`);
     // shelljs exec('', {async: true}) does not emit stdout events, so we rely on good old spawn
-    const packagerProcess = spawn('yarn', ['start', '--max-workers 1'], {
-      env: process.env,
-    });
+    const packagerProcess = child_process.spawn(
+      'yarn',
+      ['start', '--max-workers 1'],
+      {
+        env: process.env,
```
Comment (oblador) [inline scripts/run-ci-e2e-tests.js:186]: This is the default, so you can delete these options afaik

File: scripts/run-ci-e2e-tests.js [hunk 3]
```diff
@@ -288,5 +315,9 @@ try {
     echo(`Killing appium ${APPIUM_PID}`);
     exec(`kill -9 ${APPIUM_PID}`);
   }
+  if (VERDACCIO_PID) {
+    echo(`Killing verdaccio ${VERDACCIO_PID}`);
+    exec(`kill -9 ${VERDACCIO_PID}`);
```
Comment (oblador) [inline scripts/run-ci-e2e-tests.js:320]: Any particular reason for doing `-9` or just copy paste from above?

File: scripts/wait-for-verdaccio.js [hunk 4]
```diff
@@ -0,0 +1,21 @@
+#!/usr/bin/env node
```
Comment (oblador) [inline scripts/wait-for-verdaccio.js:1]: You _could_ be doing `npx wait-on` instead https://github.com/jeffbski/wait-on

File: scripts/run-ci-e2e-tests.js [hunk 5]
```diff
@@ -70,6 +72,25 @@ try {
...
+    './.circleci/verdaccio/config.yml',
+  ]);
+  VERDACCIO_PID = verdaccioProcess.pid;
+  exec('npm set registry http://localhost:4873');
+
+  exec('node ./scripts/wait-for-verdaccio.js');
+
+  echo('Publish packages');
+  const packages = fs.readdirSync('packages');
+  packages.forEach(current_package => {
```
Comment (oblador) [inline scripts/run-ci-e2e-tests.js:88]: This is not safe, I think it should at a minimum check that it's a folder etc, but preferably go after the yarn workspace config. You  can access it through `yarn workspaces info`

File: .circleci/verdaccio/config.yml [hunk 6]
```diff
@@ -0,0 +1,27 @@
+storage: ./storage
+auth:
+  htpasswd:
+    file: ./htpasswd
+uplinks:
+  npmjs:
+    url: https://registry.npmjs.org/
```
Comment (oblador) [inline .circleci/verdaccio/config.yml:7]: Not sure if it matters, but the lockfile seems to be using predominantly `https://registry.yarnpkg.com`, which I guess is also just a npmjs proxy.

File: .circleci/verdaccio/config.yml [hunk 7]
```diff
@@ -0,0 +1,27 @@
+storage: ./storage
+auth:
+  htpasswd:
+    file: ./htpasswd
```
Comment (oblador) [inline .circleci/verdaccio/config.yml:4]: Couldn't find this file in the commits, did you forget to add it or doesn't it matter?

File: scripts/run-ci-e2e-tests.js (hunk 3 above)
Comment (fortmarek) [reply scripts/run-ci-e2e-tests.js:320]: Using `-9` forces the program to be killed and yes, it's reusing the convention from the other `kill` invocations. I'd rather keep it unless there's a good reason not to.

File: .circleci/verdaccio/config.yml (hunk 7 above)
Comment (fortmarek) [reply .circleci/verdaccio/config.yml:4]: It's created by verdaccio automatically, so it should not be committed.

File: .circleci/verdaccio/config.yml [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+    max_fails: 40
+    maxage: 30m
+    timeout: 60s
+    fail_timeout: 10m
+    cache: false
+    agent_options:
+      keepAlive: true
+      maxSockets: 40
+      maxFreeSockets: 10
+packages:
```
Comment (juanpicado) [inline .circleci/verdaccio/config.yml:17]: I'd recommend optimisations like https://github.com/facebook/docusaurus/pull/7967 to avoid the registry proxy private packages to outside (publish might fails due network conditions) also help to keep isolated the packages.

File: .circleci/verdaccio/config.yml [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+  '@*/*':
+    access: $all
+    publish: $all
+    proxy: npmjs
+  '**':
+    access: $all
+    publish: $all
+    proxy: npmjs
+logs:
+  - {type: file, path: verdaccio.log, format: pretty, level: debug}
```
Comment (juanpicado) [inline .circleci/verdaccio/config.yml:27]: Debug is really verbose, I'd recommend `warn` and also `pretty` is pretty expensive in terms of execution, if you don't need the coolers or fancy arrows, you might consider `json` instead to boost the registry.

File: .circleci/verdaccio/config.yml (hunk 7 above)
Comment (juanpicado) [reply .circleci/verdaccio/config.yml:4]: An alternative in case you want to be more transparent https://github.com/verdaccio/monorepo/tree/9.x/plugins/auth-memory plugin might helps, it was made mostly for CI environments.

File: .circleci/verdaccio/config.yml (hunk 6 above)
Comment (juanpicado) [reply .circleci/verdaccio/config.yml]: I think is just a CNAME https://twitter.com/1stvamp/status/1000174791129747456 but I have no clue how looks like now.

File: .circleci/verdaccio/config.yml (hunk 7 above)
Comment (juanpicado) [reply .circleci/verdaccio/config.yml:4]: Well, if you have `[$all](publish: $all)` I think even anonymous users can just publish so this might be just decoration.

File: .circleci/verdaccio/config.yml (hunk 8 above)
Comment (fortmarek) [reply .circleci/verdaccio/config.yml]: This is perfect, thank you so much 🤞

File: template/package.json [hunk 10]
```diff
@@ -16,7 +16,7 @@
   "devDependencies": {
     "@babel/core": "^7.12.9",
     "@babel/runtime": "^7.12.5",
-    "@react-native-community/eslint-config": "^2.0.0",
+    "@react-native-community/eslint-config": "^3.0.0",
```
Comment (fortmarek) [inline template/package.json:19]: @kelset any reason why _not_ to do this? With the current setup, I only locally publish latest versions of packages in the monorepo and `@react-native-community/eslint-config` current major version is 3. Without the bump, the template installation fails since it cannot find a compatible `@react-native-community/eslint-config` version.

Comment (kelset) [reply template/package.json:19]: I think it's fine, I wonder why noone did it before

**Comment by cortinico** [comment]
Thanks for doing this @fortmarek. I haven't had the time to review this yet, but I've posted an update on the Monorepo effort here: https://github.com/facebook/react-native/issues/34692 so we're aligned on the next steps

File: .circleci/verdaccio/config.yml (hunk 7 above)
Comment (fortmarek) [reply .circleci/verdaccio/config.yml]: Given this note:
> If you want to use this piece of software, do it at your own risk. This plugin is being used for unit testing.

I'd rather keep the current setup 👀

File: .circleci/verdaccio/config.yml (hunk 6 above)
Comment (fortmarek) [reply .circleci/verdaccio/config.yml]: For consistency, I changed the name but both should work the same way afaict

File: scripts/run-ci-e2e-tests.js [hunk 11]
```diff
@@ -70,6 +71,33 @@ try {
 
   const REACT_NATIVE_PACKAGE = path.join(ROOT, 'react-native-*.tgz');
 
+  describe('Set up Verdaccio');
```
Comment (cortinico) [inline scripts/run-ci-e2e-tests.js:74]: I would move this block to a `scripts/setup-verdaccio.js` so that other scripts can reuse it.

File: scripts/run-ci-e2e-tests.js [hunk 12]
```diff
@@ -70,6 +71,33 @@ try {
...
+  exec('echo "//localhost:4873/:_authToken=secretToken" > .npmrc');
+
+  describe('Publish packages');
+  const packages = JSON.parse(
+    JSON.parse(exec('yarn --json workspaces info').stdout).data,
+  );
+  Object.keys(packages)
+    .filter(
+      packageName =>
+        packageName !== '@react-native/tester' &&
```
Comment (cortinico) [inline scripts/run-ci-e2e-tests.js:92]: Now `'@react-native/tester'` is private, so perhaps this is not needed anymore?

File: .circleci/verdaccio.yml [hunk 13]
```diff
@@ -0,0 +1,44 @@
+storage: ./storage
+auth:
+  htpasswd:
+    file: ./htpasswd
+uplinks:
+  npmjs:
+    url: https://registry.yarnpkg.com/
```
Comment (cortinico) [inline .circleci/verdaccio.yml:7]: Why is this not `https://registry.npmjs.org/`? Plus having the variable as `npmjs` while the hostname is yarnpkg is confusing

File: .circleci/verdaccio.yml [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+  'react-native-codegen':
+    access: $all
+    publish: $all
+  'react-native-gradle-plugin':
+    access: $all
+    publish: $all
+  '@*/*':
+    access: $all
+    publish: $authenticated
+    proxy: npmjs
```
Comment (cortinico) [inline .circleci/verdaccio.yml:38]: Why is this needed?

**Review by cortinico — APPROVED** [review]
Great stuff 👍

File: .circleci/verdaccio.yml (hunk 14 above)
Comment (fortmarek) [reply .circleci/verdaccio.yml:38]: This prohibits unauthenticated users to published organization-scoped packages. I've seen this being done both in the Microsoft PR and the docusaurus one. @juanpicado, do you know the exact reasoning behind this?

File: scripts/run-ci-e2e-tests.js (hunk 12 above)
Comment (fortmarek) [reply scripts/run-ci-e2e-tests.js:92]: good point, we should be able to skip that.

File: .circleci/verdaccio.yml (hunk 13 above)
Comment (fortmarek) [reply .circleci/verdaccio.yml:7]: This came about from [this](https://github.com/facebook/react-native/pull/34577#discussion_r965885227) conversation. I don't mind either way, let me revert it back to `https://registry.npmjs.org/` 👍

**Comment by fortmarek** [comment]
Thanks for the review, the comments have been addressed 🤞

File: .circleci/verdaccio.yml (hunk 14 above)
Comment (cortinico) [reply .circleci/verdaccio.yml:38]: Great thanks for the clarification 👍
